### PR TITLE
TF-2084 add/remove actions in UI

### PR DESCRIPTION
**Pull request by hieutbui**
## Issue
---
#2084 
#2086 

## Resolved 
---
- web:

https://github.com/linagora/tmail-flutter/assets/80142234/5d431bd3-98a6-4e70-9224-16bd786eba77

- android:

https://github.com/linagora/tmail-flutter/assets/80142234/e7b7abc2-e18a-4acb-b751-12e25b182cf1

### CODE REVIEW

**Comment by dab246** [comment]
Please demo on mobile

File: lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import 'package:equatable/equatable.dart';
+import 'package:tmail_ui_user/features/rules_filter_creator/presentation/model/email_rule_filter_action.dart';
+
+class RuleFilterActionArguments with EquatableMixin {
+  final EmailRuleFilterAction? action;
+  final dynamic detailed;
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart:6]: Use of `dynamic` is not recommended. You should specify a specific style for it. Furthermore, the variable name `detailed` is unclear.

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 2]
```diff
@@ -64,6 +65,12 @@ class RulesFilterCreatorController extends BaseMailboxController {
   final listRuleConditionValueArguments = RxList<RulesFilterInputFieldArguments>();
   final conditionCombinerType = Rxn<ConditionCombiner>();
 
+  final TextEditingController forwardEmailController = TextEditingController();
+  final FocusNode forwardEmailFocusNode = FocusNode();
+  final listEmailRuleFilterActionSelected = RxList<RuleFilterActionArguments>();
+  final maxEmailRuleFilterActionNumber = 5;
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:71]: why need it?. If you want to test the number of actions you can write
```
int maxCountAction = EmailRuleFilterAction.values.length;
```

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 3]
```diff
@@ -160,7 +169,13 @@ class RulesFilterCreatorController extends BaseMailboxController {
           controller: TextEditingController(),
         );
         listRuleConditionValueArguments.add(newRuleConditionValueArguments);
-        emailRuleFilterActionSelected.value = EmailRuleFilterAction.moveMessage;
+        isShowAddAction.value = true;
+        RuleFilterActionArguments newRuleFilterAction = RuleFilterActionArguments(
+          action: null,
+          detailed: null,
+          error: '',
+        );
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:177]: Should be create empty constructor  in  RuleFilterActionArguments by use factory
```suggestion

factory RuleFilterActionArguments.empty() => RuleFilterActionArguments(action: null, detailed: null, error: '' );

```

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 4]
```diff
@@ -294,11 +309,34 @@ class RulesFilterCreatorController extends BaseMailboxController {
...
   }
 
-  void selectEmailRuleFilterAction(EmailRuleFilterAction? newAction) {
-    emailRuleFilterActionSelected.value = newAction;
+  void selectEmailRuleFilterAction(EmailRuleFilterAction? newAction, int ruleFilterActionIndex) {
+    RuleFilterActionArguments newRuleFilterAction = RuleFilterActionArguments(
+      action: newAction,
+      detailed: null,
+      error: ''
+    );
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:317]: same

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 5]
```diff
@@ -435,4 +482,37 @@ class RulesFilterCreatorController extends BaseMailboxController {
       conditionCombinerType.value = combinerType;
     }
   }
+  void tapAddAction() {
+    RuleFilterActionArguments newRuleFilterAction = RuleFilterActionArguments(
+      action: null,
+      detailed: null,
+      error: '',
+    );
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:490]: same

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart [hunk 6]
```diff
@@ -640,6 +535,55 @@ class RuleFilterCreatorView extends GetWidget<RulesFilterCreatorController> {
     });
   }
 
+  Widget _buildListRuleFilterActionList(
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart:538]: Use split function NOT recommended. Should be create stateless widget

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart [hunk 7]
```diff
@@ -770,12 +716,52 @@ class RuleFilterCreatorView extends GetWidget<RulesFilterCreatorController> {
...
-        controller.selectEmailRuleFilterAction(ruleAction);
+        if (ruleAction != controller.listEmailRuleFilterActionSelected[ruleActionIndex].action) {
+          controller.selectEmailRuleFilterAction(ruleAction, ruleActionIndex);
+        }
         popBack();
       }))
     .build();
   }
 
+  Widget _buildAddActionButton(
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart:727]: same

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart [hunk 8]
```diff
@@ -770,12 +716,52 @@ class RuleFilterCreatorView extends GetWidget<RulesFilterCreatorController> {
...
+                ),
+                const SizedBox(width: 15,),
+                Text(
+                  AppLocalizations.of(context).addAction,
+                  maxLines: 1,
+                  style: const TextStyle(
+                    fontWeight: FontWeight.w500,
+                    fontSize: 17,
+                    color: AppColor.primaryColor
+                  )
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart:753]: Should be add to `style` file

File: lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import 'package:equatable/equatable.dart';
+import 'package:model/model.dart';
+import 'package:tmail_ui_user/features/rules_filter_creator/presentation/model/email_rule_filter_action.dart';
+
+class RuleFilterActionArguments with EquatableMixin {
+  final EmailRuleFilterAction? action;
+  final String? forwardEmail;
+  final PresentationMailbox? mailbox;
+  final String? error;
```
Comment (hoangdat) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart:9]: why we need `error` in this argument?

File: lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    this.forwardEmail,
+    this.mailbox,
+    this.error,
+  });
+
+  factory RuleFilterActionArguments.empty() => RuleFilterActionArguments(
+    action: null,
+    forwardEmail: null,
+    mailbox: null,
+    error: '',
```
Comment (hoangdat) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart:22]: I am not a fan of it :)

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_action_widget.dart [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                    bottomLeft: Radius.circular(RuleFilterActionStyles.mainBorderRadius),
+                    topLeft: Radius.circular(RuleFilterActionStyles.mainBorderRadius),  
+                  );
+              return Container(
+                padding: const EdgeInsets.all(RuleFilterActionStyles.mainPadding),
+                decoration: BoxDecoration(
+                  color: AppColor.colorBackgroundFieldConditionRulesFilter,
+                  borderRadius: borderRadius,
+                ),
+                child: screenType != RuleFilterConditionScreenType.mobile
```
Comment (hoangdat) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_action_widget.dart:90]: why not base on the ResponseUtils?

File: lib/features/rules_filter_creator/presentation/model/email_rule_filter_action.dart [hunk 12]
```diff
@@ -1,15 +1,45 @@
...
+      case EmailRuleFilterAction.maskAsSeen:
+        return true;
+      case EmailRuleFilterAction.startIt:
+        return true;
+      case EmailRuleFilterAction.rejectIt:
+        return true;
+      case EmailRuleFilterAction.markAsSpam:
+        return true;
+      case EmailRuleFilterAction.forwardTo:
+        return false;
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/model/email_rule_filter_action.dart:42]:
```suggestion
return this != EmailRuleFilterAction.forwardTo;
```

File: lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import 'package:equatable/equatable.dart';
+import 'package:model/model.dart';
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart:2]: Only import necessary classes

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_action_detailed_builder.dart [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+          editingController: forwardEmailEditingController,
+          focusNode: forwardEmailFocusNode,
+          onChangeAction: forwardEmailOnChangeAction,
+        );
+      case EmailRuleFilterAction.maskAsSeen:
+      case EmailRuleFilterAction.startIt:
+      case EmailRuleFilterAction.rejectIt:
+      case EmailRuleFilterAction.markAsSpam:
+      default:
+        return Container();
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_action_detailed_builder.dart:54]: Should be use `const SizeBox.shink();`

**Review by dab246 — APPROVED** [review]

File: lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import 'package:equatable/equatable.dart';
+import 'package:model/mailbox/presentation_mailbox.dart';
+import 'package:tmail_ui_user/features/rules_filter_creator/presentation/model/email_rule_filter_action.dart';
+
+class RuleFilterActionArguments with EquatableMixin {
```
Comment (hoangdat) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart:5]: please separate argument for each action. Thanks

Comment (hieutbui) [reply lib/features/rules_filter_creator/presentation/model/rule_filter_action_arguments.dart]: I separated it!

**Review by hoangdat — APPROVED** [review]

**Pull request merged**